logendpoint.cpp

```diff
@@ -48,44 +52,62 @@ void LogEndpoint::_send_msg(const mavlink_message_t *msg, int target_sysid)
     _stat.read.handled_bytes += buffer.len;
 }
 
-char *LogEndpoint::_get_filename(const char *extension)
+uint32_t LogEndpoint::_get_prefix(DIR *dir)
 {
-    char filename[PATH_MAX] = {};
+    struct dirent *ent;
+    uint32_t u, prefix = 0;
+
+    while ((ent = readdir(dir)) != nullptr) {
+        if (sscanf(ent->d_name, "%u-", &u) == 1) {
+            if (u >= prefix && u < UINT32_MAX) {
+                prefix = ++u;
+            }
+        }
+    }
+
+    return prefix;
+}
 
+int LogEndpoint::_get_file(const char *extension, char *filename, size_t filename_size)
+{
     time_t t = time(NULL);
     struct tm *timeinfo = localtime(&t);
-    int r;
-    uint16_t i;
-
-    for (i = 0; i < UINT16_MAX; i++) {
-        if (i) {
-            r = snprintf(filename, sizeof(filename), "%s/%i-%02i-%02i_%02i-%02i-%02i_%u.%s",
-                         _logs_dir, timeinfo->tm_year + 1900, timeinfo->tm_mon + 1,
-                         timeinfo->tm_mday, timeinfo->tm_hour, timeinfo->tm_min, timeinfo->tm_sec,
-                         i, extension);
-        } else {
-            r = snprintf(filename, sizeof(filename), "%s/%i-%02i-%02i_%02i-%02i-%02i.%s", _logs_dir,
-                         timeinfo->tm_year + 1900, timeinfo->tm_mon + 1, timeinfo->tm_mday,
-                         timeinfo->tm_hour, timeinfo->tm_min, timeinfo->tm_sec, extension);
-        }
+    uint32_t i;
+    int j, r;
+    DIR *dir;
+
+    dir = opendir(_logs_dir);
+    if (!dir) {
+        return log_error_errno(errno, "Could not open log dir (%m)");
+    }
+    // Close dir when leaving function.
+    std::shared_ptr<void> defer(dir, [](auto p) { closedir(p); });
```

Comment:
It could, but construction would be a bit different, after some tries I got this working:

```c++
std::unique_ptr<DIR, std::function<void(DIR*)>> defer(dir, [](auto p) { closedir(p); });
```
Worth changing? This one is more specific than using `shared_ptr`. We could create a new generic type to solve this...